lib/device.ex

```diff
@@ -0,0 +1,149 @@
+defmodule Onvif.Device do
+  alias Onvif.Device
+  alias Onvif.Discovery.Probe
+
+  @enforce_keys [:username, :password, :ip]
+
+  defstruct @enforce_keys ++
+              [
+                :auth_type,
+                :scopes,
+                :manufacturer,
+                :model,
+                :firmware_version,
+                :serial_number,
+                :hardware_id,
+                :datetime,
+                :ntp,
+                :media_service_path,
+                current_diff: 0,
+                port: 80,
+                supports_media2?: false,
+                device_service_path: "/onvif/device_service",
+                is_nvr?: false
+              ]
+
+  @spec init(Onvif.Discovery.Probe.t(), String.t(), String.t(), boolean, boolean) ::
+          {:error, any()}
+          | %Device{}
+  def init(%Probe{} = probe_match, username, password, use_https? \\ false, use_ipv6? \\ false) do
+    with {:ok, device} <-
+           device_from_probe_match(probe_match, username, password, use_https?, use_ipv6?),
+         {:ok, device_with_datetime} <- get_date_time(device),
+         {:ok, updated_device} <- guess_auth(device_with_datetime) do
+      updated_device
+      |> get_services()
+      |> set_media_service_path()
+    end
+  end
+
+  defp device_from_probe_match(%Probe{} = probe_match, username, password, use_https?, use_ipv6?) do
+    scheme = if use_https?, do: "https", else: "http"
+
+    uri =
+      probe_match.address
+      |> get_address(use_ipv6?)
+      |> URI.parse()
+      |> Map.put(:scheme, scheme)
+      |> Map.put(:userinfo, "#{username}:#{password}")
+
+    is_nvr? = if "tds:Device" in probe_match.types, do: false, else: true
+
+    device = %Device{
+      username: username,
+      password: password,
+      ip: URI.to_string(uri),
+      port: uri.port,
+      scopes: probe_match.scopes,
+      is_nvr?: is_nvr?
+    }
+
+    {:ok, device}
+  end
+
+  defp get_address(addresses, use_ipv6?) do
+    ipv6_address =
+      Enum.find(addresses, fn address ->
+        uri = URI.parse(address)
+        String.contains?(uri.host, ":")
```

Comment:
I don't know if this is a definitive way of detecting an actual ipv6 address. it might be more effective to reach for erlang

```
> ipv6_addr |> :binary.bin_to_list() |> :inet.parse_ipv6strict_address()
{:ok, {this, tuple, has, 8, values, because, it's, ipv6}}

> not_ipv6_addr |> :binary.bin_to_list() |> :inet.parse_ipv6strict_address()
{:error, :einval}
```